grails-app/services/org/pih/warehouse/invoice/InvoiceService.groovy

```diff
@@ -460,4 +468,48 @@ class InvoiceService {
         }
         return csv
     }
+
+    CSVPrinter getInvoiceItemsCsv(List<InvoiceItem> invoiceItems) {
+        CSVPrinter csv = CSVUtils.getCSVPrinter()
+        csv.printRecord(
+                "Invoice Number",
+                "Vendor Invoice Number",
+                "Vendor",
+                "Currency",
+                "Status",
+                "Buyer Organization",
+                "Invoice Type",
+                "Code",
+                "Name",
+                "Order Number",
+                "GL Account",
+                "Budget Code",
+                "Quantity",
+                "Quantity per UoM",
+                "Amount",
+                "Total Amount",
+        )
+
+        invoiceItems?.each { InvoiceItem invoiceItem ->
+            csv.printRecord(
+                    invoiceItem.invoice?.invoiceNumber,
+                    invoiceItem.invoice?.vendorInvoiceNumber,
+                    "${invoiceItem.invoice?.party?.code} ${invoiceItem.invoice?.party?.name}",
+                    invoiceItem.invoice?.currencyUom.name,
+                    invoiceItem.invoice?.status?.name(),
+                    "${invoiceItem.invoice?.partyFrom?.code} ${invoiceItem.invoice?.partyFrom?.name}",
+                    invoiceItem.invoice?.invoiceType?.code?.name() ?: "INVOICE",
+                    invoiceItem.product?.productCode ?: "all",
+                    invoiceItem?.orderAdjustment ? invoiceItem?.description : invoiceItem.product?.name,
+                    invoiceItem.order?.orderNumber,
+                    invoiceItem.glAccount?.name,
+                    invoiceItem.budgetCode?.name,
+                    invoiceItem?.quantity,
+                    invoiceItem?.quantityPerUom,
+                    invoiceItem?.unitPrice ?: 0,
+                    invoiceItem?.totalAmount ?: 0,
```

Comment:
@drodzewicz I meant you might run into N+1 in this place, where you refer to many different entities and this would run N+1 for every invoice item. So my N+1 comment from above was not directly refering to the "two queries" you made, but overall.

Reply:
@kchelstowski Yup, you're right, I also prefer to use our data services here (or just measure the performance and make a decision based on that)